### PR TITLE
Feat: Heavy Finisher Attack

### PR DESCRIPTION
**Pull request by UriMurcia**
Here is the heavy finisher attack for each of the characters Bix and Allura. This attack is going to be controlled by the combo system, which is not implemented yet. To test this attack, I think the best option is changing the BixAttack Script to trigger this attack with left click instead of the normal attack. To do this just change this: (Use GameplayFullTrial.axoltl scene)
Captura de pantalla 2023-07-20 003925
Captura de pantalla 2023-07-20 003847

Or just copy the code from here:
- BixAttackScript.h:
```c++
#pragma once

#include "Scripting\Script.h"
#include "RuntimeInclude.h"

RUNTIME_MODIFIABLE_INCLUDE;

class ModuleInput;

class ComponentAudioSource;
class ComponentTransform;
class ComponentAnimation;
class EntityDetection;
class HeavyFinisherAttack;

class PlayerManagerScript;

enum class AttackCombo
{
	IDLE,
	FIRST_ATTACK,
	SECOND_ATTACK,
	THIRD_ATTACK
};

class BixAttackScript : public Script
{

public:
	BixAttackScript();
	~BixAttackScript() override = default;

	bool GetIsDeathTouched() const;
	void SetIsDeathTouched(bool isDeathTouch);

private:
	void Start() override;
	void Update(float deltaTime) override;

	void PerformAttack();
	void CheckCombo();

	bool IsAttackAvailable() const;
	void CheckCollision() const;

	float attackCooldown;
	float lastAttackTime;

	bool isDeathTouched;


	ComponentAudioSource* audioSource;
	ComponentTransform* transform;
	ComponentAnimation* animation;
	GameObject* animationGO;
	EntityDetection* enemyDetection;
	GameObject* enemyDetectionObject;

	ModuleInput* input;

	PlayerManagerScript* playerManager;
	AttackCombo attackComboPhase;
	HeavyFinisherAttack* heavyFinisherAttack;
	GameObject* heavyFinisherAttackGO;
};

``` 

- BixAttackScript.cpp:
```c++
#include "BixAttackScript.h"

#include "Application.h"

#include "ModuleInput.h"
#include "ModuleScene.h"
#include "Scene/Scene.h"

#include "Physics/Physics.h"

#include "Components/ComponentAudioSource.h"
#include "Components/ComponentRigidBody.h"
#include "Components/ComponentTransform.h"
#include "Components/ComponentAnimation.h"
#include "Components/ComponentScript.h"

#include "../Scripts/HealthSystem.h"
#include "../Scripts/PlayerManagerScript.h"
#include "../Scripts/EntityDetection.h"
#include "../Scripts/HeavyFinisherAttack.h"

#include "GameObject/GameObject.h"

#include "Auxiliar/Audio/AudioData.h"

#include "../MathGeoLib/Include/Geometry/Ray.h"

#include "debugdraw.h"

#include <set>

#include "AxoLog.h"

REGISTERCLASS(BixAttackScript);

BixAttackScript::BixAttackScript() : Script(), attackCooldown(0.6f), lastAttackTime(0.f), audioSource(nullptr),
	input(nullptr), animation(nullptr), animationGO(nullptr), transform(nullptr), heavyFinisherAttack(nullptr), heavyFinisherAttackGO(nullptr),
	playerManager(nullptr), attackComboPhase(AttackCombo::IDLE), enemyDetection(nullptr), enemyDetectionObject(nullptr)
{
	REGISTER_FIELD(attackCooldown, float);
	REGISTER_FIELD(animationGO, GameObject*);
	REGISTER_FIELD(enemyDetectionObject, GameObject*);
	REGISTER_FIELD(heavyFinisherAttackGO, GameObject*);
}

void BixAttackScript::Start()
{
	audioSource = owner->GetComponent<ComponentAudioSource>();
	transform = owner->GetComponent<ComponentTransform>();
	if (animationGO)
	{
		animation = animationGO->GetComponent<ComponentAnimation>();
	}

	input = App->GetModule<ModuleInput>();

	audioSource->PostEvent(AUDIO::SFX::PLAYER::WEAPON::LIGHTSABER_OPEN);
	audioSource->PostEvent(AUDIO::SFX::PLAYER::WEAPON::LIGHTSABER_HUM);

	playerManager = owner->GetComponent<PlayerManagerScript>();

	enemyDetection = enemyDetectionObject->GetComponent<EntityDetection>();
	heavyFinisherAttack = heavyFinisherAttackGO->GetComponent<HeavyFinisherAttack>();
}

void BixAttackScript::Update(float deltaTime)
{
	// Provisional here until we have a way to delay a call to a function a certain time
	// This should go inside the PerformAttack() function but delay setting it to false by 2 seconds or smth like that
	animation->SetParameter("IsAttacking", false);

	CheckCombo();
}

void BixAttackScript::PerformAttack()
{
	lastAttackTime = SDL_GetTicks() / 1000.0f;

	audioSource->PostEvent(AUDIO::SFX::PLAYER::WEAPON::LIGHTSABER_SWING);

	CheckCollision();
}

void BixAttackScript::CheckCollision() const
{
	GameObject* enemyAttacked = enemyDetection->GetEnemySelected();

	if (enemyAttacked != nullptr)
	{
		HealthSystem* healthScript = enemyAttacked->GetComponent<HealthSystem>();
		float damageAttack = playerManager->GetPlayerAttack();
		if (!isDeathTouched)
		{
			//healthScript->TakeDamage(damageAttack);
			heavyFinisherAttack->PerformHeavyFinisher(enemyAttacked->GetComponent<ComponentTransform>(), transform);
			//audioSource->PostEvent(AUDIO::SFX::PLAYER::WEAPON::LIGHTSABER_CLASH);
		}
		else
		{
			//healthScript->TakeDamage(healthScript->GetMaxHealth());
			heavyFinisherAttack->PerformHeavyFinisher(enemyAttacked->GetComponent<ComponentTransform>(), transform);
			//audioSource->PostEvent(AUDIO::SFX::PLAYER::WEAPON::LIGHTSABER_CLASH);
		}
	}
}

bool BixAttackScript::IsAttackAvailable() const
{
	return (SDL_GetTicks() / 1000.0f > lastAttackTime + attackCooldown);
}

void BixAttackScript::CheckCombo()
{
	// Attack, starting the combo
	if (input->GetMouseButton(SDL_BUTTON_LEFT) == KeyState::DOWN && IsAttackAvailable())
	{
		//LOG_VERBOSE("Pressing left mouse button");

		if (animation && attackComboPhase == AttackCombo::IDLE)
		{
			attackComboPhase = AttackCombo::FIRST_ATTACK;
			animation->SetParameter("IsAttacking", true);

			PerformAttack();
		}
	}

	// Attack, continue the combo
	if (input->GetMouseButton(SDL_BUTTON_LEFT) == KeyState::REPEAT && IsAttackAvailable())
	{
		//LOG_VERBOSE("KEEP Pressing left mouse button");

		if (animation && attackComboPhase == AttackCombo::FIRST_ATTACK)
		{
			attackComboPhase = AttackCombo::SECOND_ATTACK;
			animation->SetParameter("IsAttacking_2", true);

			PerformAttack();
		}

		else if (animation && attackComboPhase == AttackCombo::SECOND_ATTACK)
		{
			attackComboPhase = AttackCombo::THIRD_ATTACK;
			animation->SetParameter("IsAttacking_3", true);

			PerformAttack();
		}

		else if (animation && attackComboPhase == AttackCombo::THIRD_ATTACK)
		{
			attackComboPhase = AttackCombo::IDLE;
			animation->SetParameter("IsAttacking", false);

			PerformAttack();
		}
	}

	// If attack could be performed but no button pressed, lose the combo streak
	else if (input->GetMouseButton(SDL_BUTTON_LEFT) == KeyState::IDLE && IsAttackAvailable())
	{
		//LOG_VERBOSE("NOT pressing left mouse button");

		attackComboPhase = AttackCombo::IDLE;
		animation->SetParameter("IsAttacking", false);
		animation->SetParameter("IsAttacking_2", false);
		animation->SetParameter("IsAttacking_3", false);
	}
}

bool BixAttackScript::GetIsDeathTouched() const
{
	return isDeathTouched;
}

void BixAttackScript::SetIsDeathTouched(bool isDeathTouched)
{
	this->isDeathTouched = isDeathTouched;
}
```

### CODE REVIEW

**Comment by UriMurcia** [comment]
I don't know why there are that many files changed, but it's only been changed the HeavyFinisherAttack

**Review by diegusmejia117 — CHANGES_REQUESTED** [review]
Great work, just check why the lightsaber show with 0 vertices(github problem?)

**Comment by markmeliaa** [comment]
I think I fixed it @UriMurcia, it was due to the revert that you did, which deleted part of the work that the scripting team did (adding vectors and such things). Reverts usually mess things up, I'd recommend using a reset or a rebase if you'd like to delete a commit you'd already pushed.

If you could check if something is off, I tried to leave everything as correct as I could but maybe I left something out.

**Comment by markmeliaa** [comment]
> I don't know why there are that many files changed, but it's only been changed the HeavyFinisherAttack

I think I fixed it @UriMurcia, it was due to the revert that you did, which deleted part of the work that the scripting team did (adding vectors and such things). Reverts usually mess things up, I'd recommend using a reset or a rebase if you'd like to delete a commit you'd already pushed.

If you could check if something is off, I tried to leave everything as correct as I could but maybe I left something out.

**Comment by markmeliaa** [comment]
> Great work, just check why the lightsaber show with 0 vertices(github problem?)

Where do you see the lightsaber with 0 vertices? It is shown correctly in the gameplay scene.

File: Game/Assets/Models/LightSaber.fbx
Comment (markmeliaa) [inline Game/Assets/Models/LightSaber.fbx:1]: As Diego said, the lightsaber is not correctly shown when thrown, idk what could be happening, if you could take a look.

File: Game/Scripts/HeavyFinisherAttack.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+public:
+	HeavyFinisherAttack();
+	~HeavyFinisherAttack() override = default;
+
+	void PerformHeavyFinisher(ComponentTransform* target, ComponentTransform* attackOwner);
+
+private:
+	void Start() override;
+	void Update(float deltaTime) override;
```
Comment (markmeliaa) [inline Game/Scripts/HeavyFinisherAttack.h:26]: These two methods should be public, as well as the constructor and the destructor of the class.

**Review by markmeliaa — CHANGES_REQUESTED** [review]
I tried its functionality and besides the lightsaber not being thrown correctly everything seems kind of fine, my only problem would be that there are times in which I press the attack button but the heavy finisher does not get executed, do you know why that may be?

**Comment by UriMurcia** [comment]
Hi @markmeliaa, thanks a lot for your feedback! :) 
We've been looking at these two problems with @diegusmejia117 and @AuxPort66. First, I don't know why the lightsaber is not showing, it worked for me all the time, maybe was something related with the shitty merge I did(?). Also, the problem with not being able to attack, we don't why it is, I think it's something related with the OnTriggerEnter, I've seen that sometimes it does it and sometimes not :/

**Comment by markmeliaa** [comment]
> Hi @markmeliaa, thanks a lot for your feedback! :) We've been looking at these two problems with @diegusmejia117 and @AuxPort66. First, I don't know why the lightsaber is not showing, it worked for me all the time, maybe was something related with the shitty merge I did(?). Also, the problem with not being able to attack, we don't know why it is, I think it's something related with the OnTriggerEnter, I've seen that sometimes it does it correctly and sometimes not :/

Regarding the `OnTriggerEnter` that does not check the collisions correctly, we will talk with physics to see if anyone could check what is going wrong with collisions.
Regarding the lightsaber not showing, I don't think it's related to the revert because that was completely undone and restored to how the branch was before it. You say that in you machine the lightsaber is shown correctly? Do you have any other computer in which you could try to run the project?

**Comment by markmeliaa** [comment]
Seems like... the revert... got pushed again smh 😥 @UriMurcia

**Pull request closed without merging**